doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,125 @@
+
+##### Effect
+
+The pods will go in an unknown/failed state and eventually be removed from the API server, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+
+#### Reasons
+
+Currently etcd operator handles cases A and B but when a cluster becomes unquorate (cases C and D) the current option is to restore the cluster from a backup.
+
+User may desire consistency over availability and prefer to wait for the cluster the return in a quorate state instead of restoring from a backup (that will contain old data).
+
+This proposal aims to fix cases C and D permitting to the cluster the return in a quorate state without the need to restore from a backup.
+
+## Proposed Changes
+
+#### Changes
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a new member also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume claims cleanup:
+  - set on the PVC an owner like already done with other resources.
+  - in the reconcile loop, remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API
+  - garbage collect also PVCs
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so their names be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management
+
+#### Problems
+
+This first part won't fix case C and D.
+
+### Part 2
+
+#### Changes
+
+- Change reconcile logic to handle deleted pods that had a persistent volume:
+
+If a pod of an etcd cluster member is deleted and it had a persistent volume claim associated, instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim. The pod recreation will happen only if the pod doesn't exists anymore in the API because we'll keep the same pod name. This also enable a MANDATORY property: at most once pod existence logic to avoid having in the API at the same time two pods with associated the same PVC.
+
+#### Results
+
+With this change if a majority of pods are deleted the etcd-operator will recreate them keeping their previous data, so case C and D will be fixed.
+
+
+#### Problems
+
+- until the pod isn't removed from the api server it won't be recreated (for some more details see the [Pod Safety, Consistency Guarantees, and Storage Implications](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/pod-safety.md) k8s design doc)
```

Comment:
> until the pod isn't removed from the api server

We could not and would not solve this problem here.

Also, the proposal is about some limitations we need to understand and future directions heading to. It shouldn't be the problems in here. Reference is enough. It should be more practical problems to be written down.

Reply:
I should move this out the "problem" section or just remove it since it's just a note on how k8s works. I mentioned it because the below proposed enhancements avoids this case.

Reply:
Just remove it.

To sync up: We can't have at-most-once guarantee of pods. But we can have at most once guarantee of etcd pod running because underlying PV (GCE PD, EBS for instance) guarantees only one binding. Probably this should be mentioned.

I'm not really sure about the "proposed enhancements". Isn't it the same as the solutions to part 2?

Reply:
> We can't have at-most-once guarantee of pods

I lost you :stuck_out_tongue_closed_eyes:.
I think we can have and must have (to avoid some volumes kind like iscsi being mounted on more than one node/pod at a time) at-most-once guarantee of pods. This is what stateful sets do and it's an enforced behavior of all other custom sf pod controller by the linked design doc. 

Relying on underlying storage that can be mounted only on one node at a time is an automatic fix to avoid data corruptions but won't work for all kind of storage, that's one of the reasons why that proposal has been created.

Note that this is already implemented in the k8s core components. The implementation is quite easy on the controller side: just use the same pod name so you can recreate it only if it's gone from the API. The core is in the node controller and how a pod is being deleted (no default force deletion in kubectl): 
https://github.com/kubernetes/kubernetes/pull/35235
https://github.com/kubernetes/kubernetes/pull/35484

As a side note, that the reason why there's a need, on every env (cloud or baremetal) to manually fence partitioned nodes or the implementation of a fence-controller.

Reply:
> I'm not really sure about the "proposed enhancements". Isn't it the same as the solutions to part 2?

The subtle difference is in the choices we can have when a member pod having data on a PV is gone:

1. if the cluster is quorate we have 2 choices:
 * recreate the pod of that member
 * delete and add a new member
2. if the cluster is not quorate my only choice is to recreate the pod of that member

Part 2 proposes to always recreate the pod of the member regardless if the cluster is quorate or not.
But this will fail forever in the case of corrupted data or other failures on starting the etcd member process.

So the enhancement says to recreate pods only if the cluster isn't quorate and when quorate use the remove/add member logic.

Perhaps considering these cases under the "tolerations" point of view is a better approach than this "fixed" logic